fastlane/completion.zsh

```diff
@@ -0,0 +1,36 @@
+#!/bin/zsh
+
+# This is a revised version of the ZSH completion script for fastlane
+# which adds support for some lanes being defined in `fastlane/lanes/*.rb`
+#
+# Based on https://github.com/fastlane/fastlane/blob/master/fastlane/lib/assets/completions/completion.zsh
+#
+# To install, copy this file to `~/.fastlane/completions/completion.zsh`,
+# then `source` it from your `~/.zshrc`.
+#
+_fastlane_complete() {
+  local word completions file
+  word="$1"
+
+  # look for Fastfile either in this directory or fastlane/ then grab the lane names
+  if [[ -e "Fastfile" ]] then
+    file=("Fastfile")
+  elif [[ -e "fastlane/Fastfile" ]] then
+    file=("fastlane/Fastfile")
+  elif [[ -e ".fastlane/Fastfile" ]] then
+    file=(".fastlane/Fastfile")
+  else
+    return 1
+  fi
+
+  # Add any fastlane/lanes/*.rb file too, if any
+  files=("$file" fastlane/lanes/*.rb(N))
```

Comment:
Tip: `(N)` at the end of the glob pattern toggles the `nullglob` mode just for that pattern (similar to doing `set -o nullglob` except that `(N)` is only local to the pattern while `set -o nullglob` would set the flag globally on the whole shell, while we want to avoid changing the global settings for the user under their nose). `(N)` (aka nullglob`) basically means: if there was no match for this glob pattern (i.e. you're in a repo where there are not files matching `fastlane/lanes/*.rb`) then don't throw an error (which would otherwise be the default), but instead just return an empty list.

Reply:
TIL